Add token substitution for assets (#2420)
sensu/sensu-docs: content/sensu-go/5.20/reference/tokens.md

Invalid templates or unmatched tokens return an error, which is logged and sent to the Sensu backend message transport.
Checks with token-matching errors are not executed.

> Token substitution is supported for [check][7], [hook][8], [handler][10], and [mutator][11] `command` attributes.

Reviewer: Token substitution is actually not supported for handlers & mutators; we now support token substitution for assets, which handlers & mutators can rely on, via the runtime_assets.

Here's a concrete example. Let's say I have this asset definition:

```
type: Asset
api_version: core/v2
metadata:
  name: sensu-go-hello-world
  namespace: default
spec:
  builds:
  - sha512: 06665fea5b7c7ee15e7822820aa7ddb791cc6338e38444e976e601bc7d7970592e806a7b88733690a238b7325437d31f85e98ae2fe47b008ca09c86530da9600
    url: "{{ .labels.asset_url }}/sensu-go-hello-world-0.0.1.tar.gz"
```

Now, checks & hooks can use the `sensu-go-hello-world` asset as a runtime_asset, and we will do token substitution with the agent's entity. The same things applies for assets referred by mutators & handlers, except we will use the backend's entity and not the agent's entity. The use case here is that you could host your assets on different URLs depending on the datacenter etc. without having to duplicate your assets.

Author: Thank you -- this is really helpful. I will update the PR text and hopefully get it right.

> Token substitution is supported for [check][7], [hook][8], [handler][10], and [mutator][11] `command` attributes.

Author: ```suggestion
Token substitution is supported for [check definition][7] `command` attributes and [hook][8] `command` attributes.
```

Only [entity attributes][4] are available for substitution.
Token substitution is not available for event filters because filters already have access to the entity.

Available attributes will always have [string values][9], such as labels and annotations.

## Manage entity labels

You can use token substitution with any defined [entity attributes][4], including custom labels.
See the [entity reference][6] for information about managing entity labels for proxy entities and agent entities.

## Manage asset changes

Use token substitution to retrieve changes to assets without manually updating your check, hook, handler, and mutator commands.

> For example, if an asset changes, the asset's specific path on disk will change.

Reviewer: Same as ^. Not sure if we should document the asset local path in a subsequent PR once we have the actual implementation.

Author: yes, I think you are right if that will not be part of 5.20.

> Use token substitution to retrieve changes to assets without manually updating your check, hook, handler, and mutator commands.

Author: ```suggestion
Use token substitution in your [asset][12] definitions, which your [checks][7], [hooks][8], [handlers][10] and [mutators][11] can use via their `runtime_assets` attributes.
Token substitution in assets allows you to host your assets on different URLs (for example, at different datacenters) without duplicating your assets.
```

> Token substitution is supported for [check definition][7] `command` attributes and [hook][8] `command` attributes.

Author: ```suggestion
Token substitution is supported for [check definition][7] `command` attributes and [hook][8] `command` attributes, as well as for [assets][12], which [handlers][10] and [mutators][11] can use via their `runtime_assets` attributes.
```

[7]: ../checks/#check-commands
[8]: ../hooks/
[9]: #token-data-type-limitations
[10]: ../handlers/

> [11]: ../mutators/

Author: ```suggestion
[11]: ../mutators/
[12]: ../assets/
```
